### PR TITLE
Only set is encrypted when encryption is enabled

### PR DESCRIPTION
**Pull request by nickvergessen**
Smashbox shareDir fails since https://github.com/owncloud/core/pull/17540

The reason is we check the file cache now and the file cache is set to "file is encrypted" when a file is uploaded to a mount point, when the file lives in a mount point.
### TODO
- [x] Add unit test
- [x] Add update step to fix all files when encryption is off and oC === 8.1.0
- [x] Fill placeholder issue with details: https://github.com/owncloud/core/issues/17846
### Steps to reproduce

**Encryption must be DISABLED**
1. Share a folder
2. As recipient edit a file in the folder
3. As owner open the edited file
### Testing matrix

| Instance | Reset | Test |
| --- | --- | --- |
| Install of 8.1.0 | :+1: | :white_check_mark: |
| Install of 8.1.0 + Enable Enc | :no_entry: | :white_check_mark: |
| Install of 8.0 + Update to 8.1.0 + Update to branch | :+1: | :white_check_mark: |
| Install of 8.0 + Enable Enc + Update to 8.1.0 + Update to branch | :warning: Log | :white_check_mark: |
| Install of 8.0 + Enable Enc + Update to 8.1.0 + Enable new Encryption  + Update to branch | :no_entry: | :white_check_mark: |
| Install of 8.0 + Enable/Disable Enc + Update to 8.1.0 + Update to branch | :no_entry: | :white_check_mark: |
| Install of 8.0 + Enable/Disable Enc + Decrypt all + Update to 8.1.0 + Update to branch | :+1: | :white_check_mark: |

@schiesbn please help with tests
cc @DeepDiver1975 thats why the smashbox tests fail on jenkins without enc and locking

### CODE REVIEW

**Comment by nickvergessen** [comment]
@DeepDiver1975 where do update files for core go?

**Comment by nickvergessen** [comment]
@rperezb please have a look, this needs testing before 8.1.1 goes out (especially the update script)

File: apps/files/appinfo/update.php [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+if ($installedVersion === '1.1.9' && (
+		// 8.1.0.x
+		($ocVersion[0] === 8 && $ocVersion[1] === 1 && $ocVersion[2] === 0)
+		||
+		// < 8.2.0.3
+		($ocVersion[0] === 8 && $ocVersion[1] === 2 && $ocVersion[2] === 0 && $ocVersion[3] < 3)
+	)) {
+
+	// Encryption is not enabled
+	if (!\OC::$server->getEncryptionManager()->isEnabled()) {
```
Comment (nickvergessen) [inline apps/files/appinfo/update.php]: Is this enough @schiesbn 
Or should we check the config for the value and appconfig for `encryption` app?

**Comment by nickvergessen** [comment]
@schiesbn can you confirm the following testing matrix:

### Testing matrix

Instance | Reset | Test
-------- | ----- | ----
Install of 8.1.0 | :+1: | :white_check_mark: 
Install of 8.1.0 + Enable Enc | :no_entry: |
Install of 8.0 | :+1: |
Install of 8.0 + Enable Enc | :no_entry: |
Install of 8.0 + Enable/Disable Enc | :no_entry: |
Install of 8.0 + Enable/Disable Enc + Decrypt all | :+1: |

File: apps/files/appinfo/update.php (hunk 1 above)
Comment (schiesbn) [inline apps/files/appinfo/update.php]: isEnabled() already checks it: https://github.com/owncloud/core/blob/master/lib/private/encryption/manager.php#L78

**Comment by nickvergessen** [comment]
@owncloud-bot retest this please

**Comment by schiesbn** [comment]
> @schiesbn can you confirm the following testing matrix

Test matrix looks good

**Comment by PVince81** [comment]
I wasn't able to reproduce the original issue with shared folders.

However the only case where I could reproduce it is as follows:
1. Leave encryption **disabled**
2. Mount an SFTP external storage under "/sftp"
3. With cadaver: "put test.txt"
4. With cadaver: "cp test.txt sftp/test.txt".

This does a Webdav COPY and uses the `copyFromStorage` code path.
In this case the target file "sftp/test.txt" receives the bogus encrypted flag.

I'll check if the migration fixes that.

**Comment by PVince81** [comment]
:+1: migration fixed the flag and on this branch the flag is not set any more.
After enabling encryption, the flag is still properly set on regular and external storage.

File: apps/files/appinfo/update.php [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+ *
+ * If this is not the case, we go with "better save then sorry" and don't change
+ * the flag but write a message to the ownCloud log file.
+ */
+
+/**
+ * @param \OCP\IDBConnection $conn
+ */
+function owncloud_reset_encrypted_flag(\OCP\IDBConnection $conn) {
+	$conn->executeUpdate('UPDATE `*PREFIX*filecache` SET `encrypted` = 0');
```
Comment (DeepDiver1975) [inline apps/files/appinfo/update.php]: can we add a condition to this to reduce the record sets being touched?

File: apps/files/appinfo/update.php [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+	// Encryption is not enabled
+	if (!\OC::$server->getEncryptionManager()->isEnabled()) {
+		$conn = \OC::$server->getDatabaseConnection();
+
+		// Old encryption is not known in app config
+		$oldEncryption = \OC::$server->getConfig()->getAppKeys('files_encryption');
+		if (empty($oldEncryption)) {
+			owncloud_reset_encrypted_flag($conn);
+		} else {
+			$query = $conn->prepare('SELECT * FROM `*PREFIX*filecache` WHERE `encrypted` = 1 AND `unencrypted_size` > 0', 1);
```
Comment (DeepDiver1975) [inline apps/files/appinfo/update.php]: use select count - no need to get all the record sets

Comment (schiesbn) [inline apps/files/appinfo/update.php]: we just get the first one (see limit), think this should be OK

**Comment by schiesbn** [comment]
added a additional condition to update only db entries with encrypted=1.

@DeepDiver1975 Please have a final look I would like to merge & backport before RC1 so that it get tested by QA and some other people before we release it.

**Comment by schiesbn** [comment]
@owncloud-bot retest this please

File: apps/files/appinfo/update.php (hunk 3 above)
Comment (DeepDiver1975) [inline apps/files/appinfo/update.php]: Ah - there is a limit involved - hard to see :see_no_evil:

**Comment by DeepDiver1975** [comment]
:+1:

**Comment by DeepDiver1975** [comment]
I'll create the backport pr ....

**Comment by schiesbn** [comment]
Thanks for taking care about the backport!

**Pull request merged**